donut/modules/auth/routes.py

```diff
@@ -12,13 +13,52 @@ def login():
 @blueprint.route('/login/submit', methods=['POST'])
 def login_submit():
     """Handle authentication."""
-    username = flask.request.form.get('username', None)
+    mask_and_user = flask.request.form.get('username', None)
     password = flask.request.form.get('password', None)
 
+    # Need to take care of the case when masquerading occurs.
+    username = None
+    mask = None
+
+    if mask_and_user is not None:
+        # Check for the character to split.
+        if ':' in mask_and_user:
+            # If it's there, we assume the second part is a mask name.
+            split = mask_and_user.split(':')
+
+            if len(split) > 2:
+                flask.flash('You cannot specify multiple masks at once.')
+                return flask.redirect(flask.url_for('auth.login'))
+
+            username, mask = split[0], split[1]
+
+            # But check to make sure the mask exists.
+            if not auth_utils.get_user_id(mask):
+                flask.flash('That person does not exist to mask.')
+                return flask.redirect(flask.url_for('auth.login'))
+        else:
+            username = mask_and_user
+
     if username is not None and password is not None:
         user_id = helpers.authenticate(username, password)
         if user_id is not None:
-            flask.session['username'] = username
+            permissions = auth_utils.get_permissions(username)
```

Comment:
Why do we need this?

Reply:
In theory, masquerading should only be necessary for debugging from a particular user's perspective; ideally, we should be able to understand everything from their point of view, including permissions they currently have. If something needs to be modified, we have power ourselves to do that without needing to transfer admin privileges while masquerading.